library/core/src/future/async_drop.rs

```diff
@@ -0,0 +1,269 @@
+}
+
+impl<T> Future for DeferredAsyncDrop<T> {
+    type Output = ();
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<()> {
+        // SAFETY: We never move out possibly immovable objects (Running::dtor)
+        let this = unsafe { self.get_unchecked_mut() };
+        if let DeferredAsyncDrop::Init { ptr, _pinned: _ } = *this {
+            // SAFETY: Guaranteed to be safe by [`deferred_async_drop`]'s
+            //   safety requirements
+            let dtor = unsafe { async_drop_in_place(ptr.as_ptr()) };
+            *this = DeferredAsyncDrop::Running { dtor };
+        }
+
+        let DeferredAsyncDrop::Running { dtor } = this else { unreachable!() };
+        // SAFETY: forwarding `poll` to the destructor
+        unsafe { Pin::new_unchecked(dtor) }.poll(cx)
+    }
+}
+
+/// Same as [`async_drop_in_place`] but with a deferred call to the
+/// `AsyncDrop::async_drop`.
+///
+/// # Safety
+///
+/// Same as [`async_drop_in_place`], but creation of the pinned mutable
+/// reference is deferred until first [`DeferredAsyncDrop::poll`] call.
+#[lang = "deferred_async_drop_ctor"]
+const unsafe fn deferred_async_drop<T>(item: *mut T) -> DeferredAsyncDrop<T> {
+    DeferredAsyncDrop::Init {
+        // SAFETY: Guaranteed by current function's safety requirements
+        ptr: unsafe { ptr::NonNull::new_unchecked(item) },
+        _pinned: PhantomPinned,
+    }
+}
+
+/// Awaits the `F1` future and then the `F2`.
+#[lang = "future_chain"]
+struct Chain<F1, F2> {
+    first: Option<F1>,
+    second: F2,
+}
+
+impl<F1, F2> Future for Chain<F1, F2>
+where
+    F1: Future<Output = ()>,
+    F2: Future<Output = ()>,
+{
+    type Output = ();
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<()> {
+        // SAFETY: We do not move any possibly immovable object from self
+        let this = unsafe { self.get_unchecked_mut() };
+        loop {
+            match &mut this.first {
+                Some(first) => {
+                    // SAFETY: Just forwarding poll call to the `self.first`
+                    ready!(unsafe { Pin::new_unchecked(first) }.poll(cx));
+                    // It might be important to destroy the first future
+                    // so that it wouldn't possibly hold any mutable
+                    // reference long enough to cause aliasing problems.
+                    this.first = None;
+                }
+                None => {
+                    // SAFETY: Just forwarding poll call to the `self.second`
+                    return unsafe { Pin::new_unchecked(&mut this.second) }.poll(cx);
+                }
+            }
+        }
+    }
+}
+
+#[lang = "future_chain_ctor"]
+const fn chain<F1, F2>(first: F1, second: F2) -> Chain<F1, F2> {
+    Chain { first: Some(first), second }
+}
+
+/// Used for trivial cases like scalar types and to guarantee idempotency
+/// of awaiting some async destructor.
+#[lang = "future_ready_unit"]
+struct ReadyUnit;
```

Comment:
we already have https://doc.rust-lang.org/std/future/struct.Ready.html, maybe make that a lang item instead? Passing a `tcx.types.unit` as the argument should not make this more complicated

Reply:
Idempotence.

Reply:
Please elaborate. I don't understand

Reply:
So if you call `std::future::Ready::poll` twice on the same object, [it's gonna panic](https://doc.rust-lang.org/src/core/future/ready.rs.html#22-24) which would be bad for potentially implicit calls during unwinding. Instead I want to guarantee async_drop_in_place futures to return `Poll::Ready(())` on polls once after first `Poll::Ready(())` was returned (I call it the fused future).

Reply:
Changed in 6bf5a695f20ed25c8297e312aeb2aa5dc9b10580